packages/core/src/root.ts

```diff
@@ -528,39 +528,65 @@ function updateProps(
 // it instead calls our wrapper around the function, which can refer to, and call, the
 // most recently-applied implementation, instead of directly calling the old implementation.
 
+type HotSwapResult = [any, HotSwapRecord[]?];
+
 function tryHotSwappingValues(
   currentValue: unknown,
   newValue: unknown,
-): [any, HotSwapRecord[]?] {
+  seen: WeakMap<any, HotSwapResult> = new WeakMap(),
+): HotSwapResult {
+  const seenValue = seen.get(currentValue);
+
+  if (seenValue) return seenValue;
+
   if (
     typeof currentValue === 'function' &&
     FUNCTION_CURRENT_IMPLEMENTATION_KEY in currentValue
   ) {
-    return [
+    const result: HotSwapResult = [
       typeof newValue === 'function' ? IGNORE : makeValueHotSwappable(newValue),
       [[currentValue as HotSwappableFunction<any>, newValue]],
     ];
+
+    seen.set(currentValue, result);
+
+    return result;
   }
 
   if (Array.isArray(currentValue)) {
-    return tryHotSwappingArrayValues(currentValue, newValue);
+    seen.set(currentValue, [IGNORE]);
+
+    const result = tryHotSwappingArrayValues(currentValue, newValue, seen);
+
+    seen.set(currentValue, result);
+
+    return result;
   }
 
-  if (
-    typeof currentValue === 'object' &&
-    currentValue != null &&
-    !isRemoteFragment(currentValue)
-  ) {
-    return tryHotSwappingObjectValues(currentValue, newValue);
+  if (isBasicObject(currentValue) && !isRemoteFragment(currentValue)) {
+    seen.set(currentValue, [IGNORE]);
+
+    const result = tryHotSwappingObjectValues(currentValue, newValue, seen);
+
+    seen.set(currentValue, result);
+
+    return result;
   }
 
   return [currentValue === newValue ? IGNORE : newValue];
```

Comment:
For this and most of the others, I did not do this because `seen` is either a `WeakMap` or `WeakSet`, which can't accept primitive types like strings or numbers. I could change these to be regular `Map`s and `Set`s, but then we store lots of values in memory unnecessarily.

Reply:
I see, thanks for clarifying. I will resolve these comments